### PR TITLE
BUG fix: structured controllers

### LINKED ISSUE #70: BUG: P and PI PTOs for multiple DOFs
This might be broken for multiple DOFs due to wrong reshaping.

```python
[File ~/Documents/WecOptTool/WecOptTool/wecopttool/pto.py:333, in ProportionalPTO.force(self, wec, x_wec, x_opt, nsubsteps)
    ]()[330]()[ def force(self, wec: WEC, x_wec: npt.ArrayLike, x_opt: npt.ArrayLike,
    ]()[331]()[           nsubsteps: int = 1) -> np.ndarray:
    ]()[332]()[     vel_td = self.velocity(wec, x_wec, x_opt, nsubsteps)
--> ]()[333]()[     force_td = np.reshape(x_opt, [-1, 1]) * vel_td
    ]()[334]()     return force_td

ValueError: operands could not be broadcast together with shapes (2,1) (101,2)
```

### LINKED ISSUE #73: Bounds for decision variable
In certain cases, the user will want to set bounds for the decision variable elements. For example, when using a proportional damping controller the gain should always be negative (or positive, depending on the convention).

To accomplish this, we can take any additional keyword arguments for [`wec.solve`](https://github.com/SNL-WaterPower/WecOptTool/blob/7d9bbc70f7cac23b468d5b073d6da9727c68be9e/wecopttool/core.py#L611-L781) and pass those along to [`scipy.optimize.minimize`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html#scipy.optimize.minimize). This would also enable other more detail control over [`scipy.optimize.minimize`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html#scipy.optimize.minimize) in the future.

### PR DESCRIPTION
**Pull request by ryancoe**
## Description
Fixes #70
Fixes #73 

## Type of PR
- [x] Bug fix

## Checklist for PR
- [x] Authors read the [contribution guidelines](https://github.com/SNL-WaterPower/WecOptTool/blob/main/.github/CONTRIBUTING.md)
- [x] The pull request is from an issue branch (not main) on your fork to the [main branch in WecOptTool](https://github.com/SNL-WaterPower/WecOptTool). 
- [x] The authors have given the admins edit access
- [x] All changes adhere to the style guide including PEP8, Docstrings, and Type Hints.
- [x] Added note in [Changelog](https://github.com/SNL-WaterPower/WecOptTool/blob/main/CHANGES.md)
~~Modified the documentation if applicable~~
- [x] Modified or added a new test
- [x] All tests pass
- [x] Documentation builds
~~[Reference or close any relevant issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)~~

### CODE REVIEW

**Comment by ryancoe** [comment]
@cmichelenstrofer - I believe we could merge this as is. The alternative would be to alter the `test_multiple_dof_fixed_structure_P` and `test_multiple_dof_fixed_structure_PI` tests to not call `solve`, but just check the `P` and `PI` method outputs with random `x_wec` and `x_opt` arrays. Let me know what you think.

File: tests/test_wecopttool.py [hunk 1]
```diff
@@ -449,6 +458,122 @@ def test_multiple_dof(regular_wave):
...
+    options = {'maxiter': 250, 'ftol': 1e-8}
+
+    _, _, x_wec, x_opt, _, _ = surge_heave_wavebot.solve(
+        regular_wave, obj_fun, nstate_opt, 
+        optim_options=options,
+        scale_x_wec=1, 
+        scale_x_opt=1e-2, 
+        scale_obj=1e-2)
+    
+    print(x_opt)
```
Comment (cmichelenstrofer) [inline tests/test_wecopttool.py:567]: probably should't have ``print`` statements in the tests. If there is something specific to check make it an assertion.

File: tests/test_wecopttool.py [hunk 2]
```diff
@@ -449,6 +458,122 @@ def test_multiple_dof(regular_wave):
...
+    # solve
+    _, _, x_wec, x_opt, _, _ = surge_heave_wavebot.solve(
+        regular_wave, obj_fun, nstate_opt, 
+        optim_options=options,
+        scale_x_wec=1e2, 
+        scale_x_opt=1e-3, 
+        scale_obj=1e-2,
+        bounds=bounds)
+    
+    print(x_opt)
```
Comment (cmichelenstrofer) [inline tests/test_wecopttool.py:537]: See comment below on the next ``print``

File: wecopttool/core.py [hunk 3]
```diff
@@ -620,6 +620,7 @@ def solve(self,
               optim_options: dict[str, Any] = {},
               use_grad: bool = True,
               maximize: bool = False,
+              **kwargs,
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:623]: I would prefer to go through the ``scipy.minimize`` API and decide which parts of it to expose to the users and pass those specifically. We already specifically give them the option to specify ``options``, ``x0``, ``constraints``, and ``fun``. Some of the other inputs they really should not specify directly. My preference would be for now to include ``bounds`` in here and later on have a closer look at the [options](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html) and decide what else to expose to users. What do you think?

File: wecopttool/core.py [hunk 4]
```diff
@@ -665,6 +666,8 @@ def solve(self,
         maximize: bool
             Whether to maximize the objective function. The default is
             ``False`` to minimize the objective function.
+        kwargs: dict
+            Passed along to scipy.minimize.solve.
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:670]: If changed to `bounds` change here too

File: wecopttool/core.py [hunk 5]
```diff
@@ -755,7 +758,7 @@ def callback(x):
         if use_grad:
             problem['jac'] = grad(obj_fun_scaled)
 
-        res = minimize(**problem)
+        res = minimize(**problem, **kwargs)
```
Comment (cmichelenstrofer) [inline wecopttool/core.py:761]: if change to `bounds` add `bounds` to `problem` and remove `**kwargs`.

**Review by cmichelenstrofer — APPROVED** [review]
* remove `print` statements from tests
* optionally do `bounds` rather than `**kwargs`

**Comment by cmichelenstrofer** [comment]
@ryancoe I agree, there is no reason to run the test. We can create any state [`x_wec`, `x_opt`], do post-processing on the `x_opt`, and call the PTO `force` function. Then do the exact same `assert` method we have. The WEC and PTO would still have to be built but we don't need to solve.

File: tests/test_wecopttool.py (hunk 1 above)
Comment (ryancoe) [reply tests/test_wecopttool.py]: removed w. 6547eaf27c20cef57ffabe4ceab1633f948a4a9e

File: tests/test_wecopttool.py (hunk 2 above)
Comment (ryancoe) [reply tests/test_wecopttool.py]: resolved w. 6547eaf27c20cef57ffabe4ceab1633f948a4a9e

File: wecopttool/core.py (hunk 3 above)
Comment (ryancoe) [reply wecopttool/core.py]: updated w. a8708e820c7db76a1fbcecf25e3bd1466511ea1e

File: wecopttool/core.py (hunk 5 above)
Comment (ryancoe) [reply wecopttool/core.py]: updated w. a8708e820c7db76a1fbcecf25e3bd1466511ea1e

File: wecopttool/core.py (hunk 4 above)
Comment (ryancoe) [reply wecopttool/core.py]: updated w. a8708e820c7db76a1fbcecf25e3bd1466511ea1e

**Pull request merged**